// exa-anime-scraper/src/scrapers/jkanime.js
const axios = require('axios')
const cheerio = require('cheerio')

class JkAnime {

	constructor({ base_url } = {}) {
        this.base_url = base_url || "https://jkanime.net/"
    }

	async getFromLink(link) {
	if(!link) throw new Error('Falta un parametro, no se especifico un link.')
		if(typeof link === 'string') {

			let htmlContent = await axios({ url: link })
			let $ = cheerio.load(htmlContent.data)

			let xd = $('.col-lg-6 ul li')

			let generos = []
			let numeroDeGeneros = $('.col-lg-6 ul li a').length
			let uwu = $('.col-lg-6 ul li a')

			for(var i = 0; i < numeroDeGeneros; i++) {
				generos.push(uwu[i].children[0].data)
			}

			let Synopsis = $('.anime__details__text p[rel=sinopsis]')

			let animeData = {
				title: $('h3', '.anime__details__title').text() || null,
				synopsis: Synopsis[0].children[0].data || null,
				generalInfo: {
					type: xd[0].children[1].data.replace(' ', '') || null,
					genrers: generos || null,
					languages: xd[2].children[1].data.replace(' ', '') || null,
					episodes: xd[3].children[1].data.replace(' ', '') || null,
					episodeDuration: xd[4].children[1].data.replace(' ', '').replace('.', '') || null,
					issued: xd[5].children[1].data.replace(' ', '') || null,
					status: xd[6].children[1].next.children[0].data || null,
					likes: $('.col-lg-6 span.vot').text() || null
				}
			}

			return animeData

		} else {
			throw new Error('El parametro asignado no es un string.')
		}
	}

	async search(name, { page: number }) {
		if(!name) throw new Error('Falta un parametro, no se especifico el nombre del anime.')
			if(!number) number = 1
				if(typeof name === 'string') {

					const baseUrl = this.base_url

					let htmlContent = await axios({ url: baseUrl +'buscar/'+ name.toLowerCase() +'/'+ number })
					let $ = cheerio.load(htmlContent.data)

					let results = []
					let elnegro = $('.row .col-lg-2.col-md-6 .anime__item')

					for(var i = 0; i < elnegro.length; i++) {
						results.push(await this.getFromLink(elnegro[i].children[1].attribs.href))
					}

					return results

				} else {
					throw new Error('El nombre del anime no es un string.')
				}
			
	}

}

module.exports = JkAnime
